erpnext/stock/report/stock_balance/stock_balance.py

```diff
@@ -215,66 +231,67 @@ def get_columns(filters):
 	return columns
 
 
-def get_conditions(filters):
-	conditions = ""
-	if not filters.get("from_date"):
-		frappe.throw(_("'From Date' is required"))
+def apply_conditions(query, filters):
+	sle = frappe.qb.DocType("Stock Ledger Entry")
 
-	if filters.get("to_date"):
-		conditions += " and sle.posting_date <= %s" % frappe.db.escape(filters.get("to_date"))
+	if to_date := filters.get("to_date"):
+		query = query.where(sle.posting_date <= to_date)
 	else:
 		frappe.throw(_("'To Date' is required"))
 
-	if filters.get("company"):
-		conditions += " and sle.company = %s" % frappe.db.escape(filters.get("company"))
+	if company := filters.get("company"):
+		query = query.where(sle.company == company)
 
-	if filters.get("warehouse"):
-		warehouse_details = frappe.db.get_value(
-			"Warehouse", filters.get("warehouse"), ["lft", "rgt"], as_dict=1
+	if warehouse_type := filters.get("warehouse_type") and not filters.get("warehouse"):
+		warehouse_table = frappe.qb.DocType("Warehouse")
+		query = (
+			query.join(warehouse_table)
+			.on(warehouse_table.name == sle.warehouse)
+			.where(warehouse_table.warehouse_type == warehouse_type)
 		)
-		if warehouse_details:
-			conditions += (
-				" and exists (select name from `tabWarehouse` wh \
-				where wh.lft >= %s and wh.rgt <= %s and sle.warehouse = wh.name)"
-				% (warehouse_details.lft, warehouse_details.rgt)
-			)
 
-	if filters.get("warehouse_type") and not filters.get("warehouse"):
-		conditions += (
-			" and exists (select name from `tabWarehouse` wh \
-			where wh.warehouse_type = '%s' and sle.warehouse = wh.name)"
-			% (filters.get("warehouse_type"))
+	if warehouse := filters.get("warehouse"):
+		children = get_descendants_of("Warehouse", warehouse, ignore_permissions=True)
+		# TODO: check perf impact
+		query = query.where(sle.warehouse.isin(children + [warehouse]))
```

Comment:
At first glance, this implementation using `isin` appears to be much faster because of removal of correlated subquery. 

TODO: check on very large database. 

Diff to move to previous `exists` query if required


```patch
diff --git a/erpnext/stock/report/stock_balance/stock_balance.py b/erpnext/stock/report/stock_balance/stock_balance.py
index d06a3992e0..3e136848f5 100644
--- a/erpnext/stock/report/stock_balance/stock_balance.py
+++ b/erpnext/stock/report/stock_balance/stock_balance.py
@@ -9,7 +9,7 @@ import frappe
 from frappe import _
 from frappe.query_builder.functions import CombineDatetime
 from frappe.utils import cint, date_diff, flt, getdate
-from frappe.utils.nestedset import get_descendants_of
+from pypika.terms import ExistsCriterion
 
 import erpnext
 from erpnext.stock.report.stock_ageing.stock_ageing import FIFOSlots, get_average_age
@@ -233,6 +233,7 @@ def get_columns(filters: StockBalanceFilter):
 
 def apply_conditions(query, filters):
 	sle = frappe.qb.DocType("Stock Ledger Entry")
+	warehouse_table = frappe.qb.DocType("Warehouse")
 
 	if to_date := filters.get("to_date"):
 		query = query.where(sle.posting_date <= to_date)
@@ -243,7 +244,6 @@ def apply_conditions(query, filters):
 		query = query.where(sle.company == company)
 
 	if warehouse_type := filters.get("warehouse_type") and not filters.get("warehouse"):
-		warehouse_table = frappe.qb.DocType("Warehouse")
 		query = (
 			query.join(warehouse_table)
 			.on(warehouse_table.name == sle.warehouse)
@@ -251,9 +251,17 @@ def apply_conditions(query, filters):
 		)
 
 	if warehouse := filters.get("warehouse"):
-		children = get_descendants_of("Warehouse", warehouse, ignore_permissions=True)
-		# TODO: check perf impact
-		query = query.where(sle.warehouse.isin(children + [warehouse]))
+		lft, rgt = frappe.db.get_value("Warehouse", warehouse, ["lft", "rgt"])
+		chilren_subquery = (
+			frappe.qb.from_(warehouse_table)
+			.select(warehouse_table.name)
+			.where(
+				(warehouse_table.lft >= lft)
+				& (warehouse_table.rgt <= rgt)
+				& (warehouse_table.name == sle.warehouse)
+			)
+		)
+		query = query.where(ExistsCriterion(chilren_subquery))
 
 	return query
 

```